torch/csrc/jit/passes/onnx.cpp

```diff
@@ -478,15 +478,18 @@ void NodeToONNX(
         onnx_registration.attr("registry")
             .attr("is_registered_op")("prim::PythonOp", opset_version)
             .cast<bool>();
+    py::bool_ is_autograd_inlining_enabled =
+        py::cast<bool>(onnx_globals.attr("GLOBALS").attr("autograd_inlining"));
     if (!py::hasattr(pyobj, "symbolic") && !is_registered_op) {
       // Inline the subgraph within the prim::PythonOp unless
       // either of these conditions are satisfied
       // 1. The torch.autograd.Function class of this node object has `symbolic`
       // method defined.
       // 2. Custom export symbolic is registered for prim::PythonOp.
-      if (operator_export_type == ::torch::onnx::OperatorExportTypes::ONNX ||
-          operator_export_type ==
-              ::torch::onnx::OperatorExportTypes::ONNX_ATEN_FALLBACK) {
+      if ((operator_export_type == ::torch::onnx::OperatorExportTypes::ONNX ||
+           operator_export_type ==
+               ::torch::onnx::OperatorExportTypes::ONNX_ATEN_FALLBACK) &&
+          (py::cast<bool>(is_autograd_inlining_enabled))) {
```

Comment:
@BowenBao This change was not part of @iK1D suggestion, but was added by the offending PR #74765.

If we don't add the code above, the repro script without autograd inlining fails with:
```bash
Traceback (most recent call last):
  File "repro88286.py", line 36, in <module>
    torch.onnx.export(
  File "<@beartype(torch.onnx.utils.export) at 0x7f60091e8ee0>", line 385, in export
  File "/opt/pytorch/torch/onnx/utils.py", line 511, in export
    _export(
  File "/opt/pytorch/torch/onnx/utils.py", line 1576, in _export
    graph, params_dict, torch_out = _model_to_graph(
  File "<@beartype(torch.onnx.utils._model_to_graph) at 0x7f6009173dc0>", line 11, in _model_to_graph
  File "/opt/pytorch/torch/onnx/utils.py", line 1134, in _model_to_graph
    graph = _optimize_graph(
  File "<@beartype(torch.onnx.utils._optimize_graph) at 0x7f6009173280>", line 90, in _optimize_graph
  File "/opt/pytorch/torch/onnx/utils.py", line 672, in _optimize_graph
    graph = _C._jit_pass_onnx(graph, operator_export_type)
RuntimeError: Dangling node reference
```

If we wrap it around `is_autograd_inlining_enabled`, export fails with
```bash
Traceback (most recent call last):
  File "repro88286.py", line 36, in <module>
    torch.onnx.export(
  File "<@beartype(torch.onnx.utils.export) at 0x7f39c11feee0>", line 385, in export
  File "/opt/pytorch/torch/onnx/utils.py", line 511, in export
    _export(
  File "/opt/pytorch/torch/onnx/utils.py", line 1615, in _export
    ) = graph._export_onnx(  # type: ignore[attr-defined]
RuntimeError: ONNX export failed: Couldn't export Python operator CheckpointFunction

Defined at:
/opt/pytorch/torch/autograd/function.py(506): apply
/opt/pytorch/torch/utils/checkpoint.py(412): checkpoint
repro88286.py(32): forward
/opt/pytorch/torch/nn/modules/module.py(1492): _slow_forward
/opt/pytorch/torch/nn/modules/module.py(1511): _call_impl
/opt/pytorch/torch/nn/modules/module.py(1502): _wrapped_call_impl
/opt/pytorch/torch/jit/_trace.py(119): wrapper
/opt/pytorch/torch/jit/_trace.py(128): forward
/opt/pytorch/torch/nn/modules/module.py(1511): _call_impl
/opt/pytorch/torch/nn/modules/module.py(1502): _wrapped_call_impl
/opt/pytorch/torch/jit/_trace.py(1269): _get_trace_graph
/opt/pytorch/torch/onnx/utils.py(910): _trace_and_get_graph_from_model
/opt/pytorch/torch/onnx/utils.py(1006): _create_jit_graph
/opt/pytorch/torch/onnx/utils.py(1130): _model_to_graph
<@beartype(torch.onnx.utils._model_to_graph) at 0x7f39c118adc0>(11): _model_to_graph
/opt/pytorch/torch/onnx/utils.py(1576): _export
/opt/pytorch/torch/onnx/utils.py(511): export
<@beartype(torch.onnx.utils.export) at 0x7f39c11feee0>(385): export
repro88286.py(36): <module>


Graph we tried to export:
graph(%prim::PythonOp_0 : Float(2, 2, strides=[2, 1], requires_grad=0, device=cpu),
      %l1.weight : Float(2, 2, strides=[2, 1], requires_grad=1, device=cpu),
      %l1.bias : Float(2, strides=[1], requires_grad=1, device=cpu)):
  %/PythonOp_output_0 : Float(2, 2, strides=[2, 1], requires_grad=0, device=cpu) = ^CheckpointFunction[inplace=0, module="torch.utils.checkpoint", onnx_name="/PythonOp"](<function B.forward.<locals>.custom.<locals>.custom_forward at 0x7f39c10e8790>, True)(%prim::PythonOp_0), scope: __main__.B:: # /opt/pytorch/torch/autograd/function.py:506:0
  %6 : Float(2, 2, strides=[2, 1], requires_grad=1, device=cpu) = onnx::Gemm[alpha=1., beta=1., transB=1, onnx_name="/l1/Gemm"](%/PythonOp_output_0, %l1.weight, %l1.bias), scope: __main__.B::/torch.nn.modules.linear.Linear::l1 # /opt/pytorch/torch/nn/modules/linear.py:114:0
  return (%6)
```

Given [CheckpointFunction](https://github.com/thiagocrepaldi/pytorch/blob/18f25fe99da92e9798ac62f9e4ba6ebdc97d62a3/torch/utils/checkpoint.py#L181) is an autograd function and the autograd inlining is disabled, this error seems to make sense.

What do you think?

Reply:
Sorry for late reply, was going to comment but was interrupted.

Your change makes sense. Another fix is to raise inside `inlineAutograd` when node has no subgraph. I think the latter provides extra fail-safe coverage so worth doing both change.

If `inlineAutograd` does not raise when there is no subgraph to inline. It won't trigger the fallback `cloneNode` solution, and will essentially do nothing. Hence you see `RuntimeError: Dangling node reference`.